swxsoc/util/schema.py

```diff
@@ -0,0 +1,1551 @@
+            )
+
+        # Derive Attributes Specific to `spectra` Data
+        if hasattr(var_data, "wcs") and getattr(var_data, "wcs") is not None:
+            spectra_attributes = self._derive_spectra_attributes(var_data)
+            measurement_attributes.update(spectra_attributes)
+
+        return measurement_attributes
+
+    def derive_time_attributes(self, data) -> OrderedDict:
+        """
+        Function to derive metadata for the time measurement.
+
+        Parameters
+        ----------
+        data : `swxsoc.swxdata.SWXData`
+            An instance of `SWXData` to derive metadata from.
+
+        Returns
+        -------
+        attributes : `OrderedDict`
+            A dict containing `key: value` pairs of time metadata attributes.
+        """
+
+        # Get the Variable Data
+        var_data = data["time"]
+        (guess_dims, guess_types, guess_elements) = self._types(var_data)
+
+        time_attributes = self.derive_measurement_attributes(
+            data, "time", guess_types=guess_types
+        )
+        # Check the Attributes that can be derived
+        time_attributes["REFERENCE_POSITION"] = self._get_reference_position(
+            guess_types[0]
+        )
+        time_attributes["RESOLUTION"] = self._get_resolution(data)
+        time_attributes["TIME_BASE"] = self._get_time_base(guess_types[0])
+        time_attributes["TIME_SCALE"] = self._get_time_scale(guess_types[0])
+        time_attributes["UNITS"] = self._get_time_units(guess_types[0])
+        return time_attributes
+
+    def derive_global_attributes(self, data) -> OrderedDict:
+        """
+        Function to derive global attributes for the given measurement data.
+
+        Parameters
+        ----------
+        data : `swxsoc.swxdata.SWXData`
+            An instance of `SWXData` to derive metadata from.
+
+        Returns
+        -------
+        attributes : `OrderedDict`
+            A dict containing `key: value` pairs of global metadata attributes.
+        """
+        global_attributes = OrderedDict()
+        # Loop through Global Attributes
+        derived_attributes = filter(
+            lambda attr_info: attr_info[1]["derived"],
+            self.global_attribute_schema.items(),
+        )
+        for attr_name, attr_schema in derived_attributes:
+            derivation_fn = getattr(self, attr_schema["derivation_fn"])
+            global_attributes[attr_name] = derivation_fn(data)
+
+        return global_attributes
+
+    def _derive_spectra_attributes(self, var_data):
+        """
+        Function to Derive WCS-Keyword Metadata Attributes for a given spectra variable
+        based on the variables `.wcs` member.
+        """
+        spectra_attributes = OrderedDict()
+
+        # WCSAXIS is a Single Attribute
+        spectra_attributes["WCSAXES"] = self._get_wcs_naxis(var_data)
+
+        # Get Sets/Collections of Attributes
+        for keyword, prop, _ in self.wcs_keyword_to_astropy_property:
```

Comment:
There is a set of attributes derived for spectra `NDCube` objects that are hardcoded and not determines based on the schema. This function would need to be overridden if you want to have a different subset of spectra-specific attributes.